### PR TITLE
Lazyppl

### PR DESCRIPTION
**Pull request by reubenharry**
Merge in the core of LazyPPL.

### CODE REVIEW

File: docs/source/usage.md [hunk 1]
```diff
@@ -474,56 +475,53 @@ hoistFirst :: (forall x. m x -> m x) -> Sequential m a -> Sequential m a
...
 
-As you can see, this is rather like `Coroutine`, except to "resume", you must provide a new `Double`, corresponding to the value of some particular random choice.
-
 Since `FreeT` is a transformer, we can use `lift` to get a `MonadSample` instance.
 
-
-A *trace* of a program of type `MonadSample m => m a` is an execution of the program, so a choice for each of the random values. Recall that `random` underlies all of the random values in a program, so a trace for a program is fully specified by a list of `Double`s, giving the value of each call to `random`.
-
-Given a probabilistic program interpreted in `FreeSampler m`, we can "run" it to produce a program in the underlying monad `m`. For simplicity, consider the case of a program `bernoulli 0.5 :: FreeSampler SamplerIO Bool`. We can then use the following function:
+`density` is then defined using the canonical property of the free monad (transformer), embodied by `iterFT`, which interprets `SamF` in the appropriate way:
```
Comment (idontgetoutmuch) [inline docs/source/usage.md]: I think the defining property of a free monad is the fold: https://hackage.haskell.org/package/free-5.1.9/docs/Control-Monad-Free.html#v:foldFree and https://hackage.haskell.org/package/free-5.1.9/docs/Control-Monad-Trans-Free.html#v:foldFreeT which means the free monad is initial (and hence can be interpreted in any desired way). TBH I don't know how this relates to `iterFT`. What does canonical mean here?

File: monad-bayes.cabal [hunk 2]
```diff
@@ -122,7 +131,7 @@ executable example
 
   if flag(dev)
     ghc-options:
-      -Wall -Wcompat -Wincomplete-record-updates
+      -threaded -O2 -Wall -Wcompat -Wincomplete-record-updates
```
Comment (idontgetoutmuch) [inline monad-bayes.cabal]: Do we really need `threaded`? Maybe at some point in the future we will want to run sequential Monte Carlo in parallel or parallel chains but I don't think we do now?

File: src/Control/Monad/Bayes/Inference/Lazy/MH.hs [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+      let t' = mutateTree p g1 t
+      -- Rerun the model with the new tree, to get a new
+      -- weight w'.
+      let (x', w') = runSampler (weighted m) t'
+      -- MH acceptance ratio. This is the probability of either
+      -- returning the new seed or the old one.
+      let ratio = w' / w
+      let (r, g2') = R.random g2
+      put g2'
+      if r < min 1 (exp $ ln ratio) -- (trace ("-- Ratio: " ++ show ratio) ratio))
```
Comment (idontgetoutmuch) [inline src/Control/Monad/Bayes/Inference/Lazy/MH.hs]: I don't think we should have commented out debug lying around. If we want the ability to run in debug mode (which is a valid requirement) then we could use the logging package: katip. But I suggest this is not the highest priority ATM.

File: src/Control/Monad/Bayes/Inference/Lazy/MH.hs [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+traceIt x = trace (show x) x
+
+b :: (MonadSample m, MonadCond m) => m [Double]
+b = do
+  x <- normal 0 1
+  factor 0.5
+  fmap (x :) (b)
+
+-- return (repeat 1)
```
Comment (idontgetoutmuch) [inline src/Control/Monad/Bayes/Inference/Lazy/MH.hs]: More commented out code

File: src/Control/Monad/Bayes/Inference/Lazy/MH.hs [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+-- run = do
+--   x <- Lazy.sampler $ Lazy.independent $ population $ normal 0 1
+--   print $ (take 2 ) $ fmap (take 2) x -- fmap (first (take 10)) x
+
+infiniteParticles :: MonadSample m => Population m a -> Population m a
+infiniteParticles = (P.fromWeightedList (return (repeat ((), 1))) >>)
+
+-- newtype Seq m a = Seq {runSeq :: P.Producer (Log Double) m a} deriving newtype (Functor, Applicative, Monad, MonadTrans)
```
Comment (idontgetoutmuch) [inline src/Control/Monad/Bayes/Inference/Lazy/MH.hs]: More commented out code

File: src/Control/Monad/Bayes/Sampler/Lazy.hs [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+sampler m = newStdGen *> (runSampler m . randomTree <$> getStdGen)
+
+independent :: Monad m => m a -> m [a]
+independent = sequence . repeat
+
+-- | 'weightedsamples' runs a probability measure and gets out a stream of (result,weight) pairs
+weightedsamples :: Weighted Sampler a -> IO [(a, Log Double)]
+weightedsamples = sampler . independent . weighted
+
+-- wiener :: Prob (Double -> State (Data.Map.Map Double Double) Double)
```
Comment (idontgetoutmuch) [inline src/Control/Monad/Bayes/Sampler/Lazy.hs]: More commented out code

**Review by idontgetoutmuch — COMMENTED** [review]
Not sure what to do about checking the actual implementation against the paper describing the algorithm.

File: docs/source/usage.md (hunk 1 above)
Comment (reubenharry) [reply docs/source/usage.md]: Updated.

**Review by idontgetoutmuch — APPROVED** [review]

**Pull request merged**